man/man1/expiry.1
'\" t
.\"     Title: expiry
.\"    Author: Julianne Frances Haugh
.\" Generator: DocBook XSL Stylesheets v1.78.0 <http://docbook.sf.net/>
.\"      Date: 10/15/2014
.\"    Manual: User Commands
.\"    Source: shadow-utils 4.2
.\"  Language: English
.\"
.TH "EXPIRY" "1" "10/15/2014" "shadow\-utils 4\&.2" "User Commands"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
expiry \- check and enforce password expiration policy
.SH "SYNOPSIS"
.HP \w'\fBexpiry\fR\ 'u
\fBexpiry\fR \fIoption\fR
.SH "DESCRIPTION"
.PP
The
\fBexpiry\fR
command checks (\fB\-c\fR) the current password expiration and forces (\fB\-f\fR) changes when required\&. It is callable as a normal user command\&.
.SH "OPTIONS"
.PP
The options which apply to the
\fBexpiry\fR
command are:
.PP
\fB\-c\fR, \fB\-\-check\fR
.RS 4
Check the password expiration of the current user\&.
.RE
.PP
\fB\-f\fR, \fB\-\-force\fR
.RS 4
Force a password change if the current user has an expired password\&.
.RE
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Display help message and exit\&.
.RE
.SH "FILES"
.PP
/etc/passwd
.RS 4
User account information\&.
.RE
.PP
/etc/shadow
.RS 4
Secure user account information\&.
.RE
.SH "SEE ALSO"
.PP

\fBpasswd\fR(5),
\fBshadow\fR(5)\&.


NEWS
$Id$

shadow-4.2 -> shadow-4.2.1					UNRELEASED

*** general
  * Properly release with a generated configure script, built manpages
  * Include the configure_userns patch from Debian

*** translation
  * Vietnamese translation update
  * French man pages translation update
  * French translation update
  * Typo fixes in German translation

shadow-4.1.5.1 -> shadow-4.2					2013-04-30

*** general
  * Handle libc whose crypt() returns NULL when passed a salt that
    violates specs or system requirements (e.g. FIPS140). This is needed
    with glibc/eglibc 2.17 for tools checking passwords (passwd (non PAM
    enabled) or newgrp), and for tools generating encrypted passwords
    (chgpasswd, chpasswd, or gpasswd when non PAM enabled or when a fixed
    crypt method is requested on the command line, and newusers, or passwd
    in their non PAM enabled versions)
  * Fix segfault when reading groups split on multiple lines. This impacts
    most user/group management tools when MAX_MEMBERS_PER_GROUP is set.

- su
  * When su receives a signal (SIGTERM, or SIGINT/SIGQUIT in non
    interactive mode), kill the child process group, rather than just the 
    immediate child.
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

- login
  * Fix segmentation faults for users without a proper home or shell in
    their passwd entries.

*** documentation
  * Fixed useradd man page (--home-dir option, instead of --home).

*** translation
  * Updated Russian translation.
  * Updated German man pages translation.
  * Fixed gshadow Japanese man page translation.

shadow-4.1.5 -> shadow-4.1.5.1					2012-05-25

- login
  * Log into utmp(x) when PAM is enabled, but do not log into wtmp.
    This complete pam_lastlog which logs into wtmp and in into utmp(x).
- su
  * non PAM enabled versions: do not fail if su is called without a
    controlling terminal.
- userdel
  * Fix segfault when userdel removes the user's group.

*** documentation
  * .so links now point to paths relative to the top-level manual hierarchy

*** translation
  * Updated French man pages translation.
  * Updated German man pages translation.
  * Updated Polish man pages translation. (logoutd.8)

shadow-4.1.4.3 -> shadow-4.1.5					2012-02-12

*** security
  * su -c could be abused by the executed command to invoke commands with
    the caller privileges. See below. (CVE-2005-4890)

*** general
  * report usage error to stderr, but report usage help to stdout (and return
    zero) when explicitly requested (e.g. with --help).
  * initial support for tcb (http://openwall.com/tcb/) for useradd,
    userdel, usermod, chage, pwck, vipw.
  * Added support for ACLs and Extended Attributes in useradd and usermod.
    Support shall be enabled with the new --with-acl or --with-attr
    configure options.
  * Added diagnosis for lock failures.
  * use libsemanage instead of the semanage tool.

- chage
  * Add --root option.
- chfn
  * Add --root option.
- chgpasswd
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- chpasswd
  * PAM enabled versions: restore the -e option to allow restoring
    passwords without knowing those passwords. Restore together the -m
    and -c options. (These options were removed in shadow-4.1.4 on PAM
    enabled versions)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry.
  * Add --root option.
- chsh
  * Add --root option.
- faillog
  * The -l, -m, -r, -t options only act on the existing users, unless -a is
    specified.
  * Add --root option.
- gpasswd
  * Add --root option.
- groupadd
  * Add --root option.
- groupdel
  * Add --root option.
- groupmems
  * Fix parsing of gshadow entries.
  * Add --root option.
- groupmod
  * Fixed groupmod when configured with --enable-account-tools-setuid.
  * When the gshadow file exists but there are no gshadow entries, an entry
    is created if the password is changed and group requires a
    shadow entry.
  * Add --root option.
- grpck
  * Add --root option.
  * NIS entries were dropped by -s (sort).
- grpconv
  * Add --root option.
- grpunconv
  * Add --root option.
- lastlog
  * Add --root option.
- login
  * Fixed limits support (non PAM enabled versions only)
  * Added support for infinite limits and group based limits (non PAM
    enabled versions only)
  * Fixed infinite loop when CONSOLE is configured with a colon-separated
    list of TTYs.
  * Fixed warning and support for CONSOLE_GROUPS for users member of more
    than 16 groups.
  * Do not log into utmp(x) or wtmp when PAM is enabled. This is done by
    pam_lastlog.
- newgrp, sg
  * Fix parsing of gshadow entries.
- newusers
  * Add --root option.
- passwd
  * Add --root option.
- pwpck
  * NIS entries were dropped by -s (sort).
  * Add --root option.
- pwconv
  * Add --root option.
- pwunconv
  * Add --root option.
- useradd
  * If the skeleton directory contained hardlinked files, copies of the
    hardlink were removed from the skeleton directory.
  * Add --root option.
- userdel
  * Check the existence of the user's mail spool before trying to remove
    it. If it does not exist, a warning is issued, but no failure.
  * Do not remove a group with the same name as the user (usergroup) if
    this group isn't the user's primary group.
  * Add --root option.
  * Add --selinux-user option.
- usermod
  * Accept options in any order (username not necessarily at the end)
  * When the shadow file exists but there are no shadow entries, an entry
    is created if the password is changed and passwd requires a
    shadow entry, or if aging features are used (-e or -f).
  * Add --root option.
- su
  * Document the su exit values.
  * When su receives a signal, wait for the child to terminate (after

src/expiry.c
/*
 * Copyright (c) 1994       , Julianne Frances Haugh
 * Copyright (c) 1996 - 2000, Marek Michałkiewicz
 * Copyright (c) 2001 - 2006, Tomasz Kłoczko
 * Copyright (c) 2007 - 2011, Nicolas François
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the copyright holders or contributors may not be used to
 *    endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT
 * HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <config.h>

#ident "$Id: expiry.c 3640 2011-11-19 21:51:52Z nekral-guest $"

#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <sys/types.h>
#include <getopt.h>
#include "defines.h"
#include "prototypes.h"
/*@-exitarg@*/
#include "exitcodes.h"

/* Global variables */
const char *Prog;
static bool cflg = false;

/* local function prototypes */
static RETSIGTYPE catch_signals (unused int sig);
static /*@noreturn@*/void usage (int status);
static void process_flags (int argc, char **argv);

/*
 * catch_signals - signal catcher
 */
static RETSIGTYPE catch_signals (unused int sig)
{
	exit (10);
}

/*
 * usage - print syntax message and exit
 */
static /*@noreturn@*/void usage (int status)
{
	FILE *usageout = (E_SUCCESS != status) ? stderr : stdout;
	(void) fprintf (usageout,
	                _("Usage: %s [options]\n"
	                  "\n"
	                  "Options:\n"),
	                Prog);
	(void) fputs (_("  -c, --check                   check the user's password expiration\n"), usageout);
	(void) fputs (_("  -f, --force                   force password change if the user's password\n"
	                "                                is expired\n"), usageout);
	(void) fputs (_("  -h, --help                    display this help message and exit\n"), usageout);
	(void) fputs ("\n", usageout);
	exit (status);
}

/*
 * process_flags - parse the command line options
 *
 *	It will not return if an error is encountered.
 */
static void process_flags (int argc, char **argv)
{
	bool fflg = false;
	int c;
	static struct option long_options[] = {
		{"check", no_argument, NULL, 'c'},
		{"force", no_argument, NULL, 'f'},
		{"help",  no_argument, NULL, 'h'},
		{NULL, 0, NULL, '\0'}
	};

	while ((c = getopt_long (argc, argv, "cfh",
	                         long_options, NULL)) != -1) {
		switch (c) {
		case 'c':
			cflg = true;
			break;
		case 'f':
			fflg = true;
			break;
		case 'h':
			usage (E_SUCCESS);
			/*@notreached@*/break;
		default:
			usage (E_USAGE);
		}
	}

	if (! (cflg || fflg)) {
		usage (E_USAGE);
	}

	if (cflg && fflg) {
		fprintf (stderr,
		         _("%s: options %s and %s conflict\n"),
		         Prog, "-c", "-f");
		usage (E_USAGE);
	}

	if (argc != optind) {
		fprintf (stderr,
		         _("%s: unexpected argument: %s\n"),
		         Prog, argv[optind]);
		usage (E_USAGE);
	}
}

/* 
 * expiry - check and enforce password expiration policy
 *
 *	expiry checks (-c) the current password expiration and forces (-f)
 *	changes when required. It is callable as a normal user command.
 */
int main (int argc, char **argv)
{
	struct passwd *pwd;
	struct spwd *spwd;

	Prog = Basename (argv[0]);

	sanitize_env ();

	/* 
	 * Start by disabling all of the keyboard signals.
	 */
	(void) signal (SIGHUP, catch_signals);
	(void) signal (SIGINT, catch_signals);
	(void) signal (SIGQUIT, catch_signals);
#ifdef	SIGTSTP
	(void) signal (SIGTSTP, catch_signals);
#endif

	/*
	 * expiry takes one of two arguments. The default action is to give
	 * the usage message.
	 */
	(void) setlocale (LC_ALL, "");
	(void) bindtextdomain (PACKAGE, LOCALEDIR);
	(void) textdomain (PACKAGE);

	OPENLOG ("expiry");

	process_flags (argc, argv);

	/*
	 * Get user entries for /etc/passwd and /etc/shadow
	 */
	pwd = get_my_pwent ();
	if (NULL == pwd) {
		fprintf (stderr, _("%s: Cannot determine your user name.\n"),
		         Prog);
		SYSLOG ((LOG_WARN, "Cannot determine the user name of the caller (UID %lu)",
		         (unsigned long) getuid ()));
		exit (10);
	}
	spwd = getspnam (pwd->pw_name); /* !USE_PAM, No need for xgetspnam */

	/*
	 * If checking accounts, use agecheck() function.
	 */
	if (cflg) {
		/*
		 * Print out number of days until expiration.
		 */
		agecheck (spwd);

		/*
		 * Exit with status indicating state of account.
		 */
		exit (isexpired (pwd, spwd));
	}

	/*
	 * Otherwise, force a password change with the expire() function.
	 * It will force the change or give a message indicating what to
	 * do.
	 * It won't return unless the account is unexpired.
	 */
	(void) expire (pwd, spwd);

	return E_SUCCESS;
}

